Make network layer compatible with GHC 8.10

mrBliss opened the pull request:
Update the dependency bounds and the code itself to be compatible with both GHC 8.6.5 and GHC 8.10.2. See the individual commits for more information about the required changes.

mrBliss commented:
This PR doesn't switch yet to GHC 8.10.2, that will eventually be done by https://github.com/input-output-hk/ouroboros-network/pull/2540.

I'd like to merge this PR ASAP, otherwise it will have tons of merge conflicts.

@coot Some network tests are failing because I replaced `fail` with `error`: https://hydra.iohk.io/build/3933045/nixlog/6 and https://hydra.iohk.io/build/3933046/nixlog/1. What's the appropriate solution?

mrBliss commented:
I have extracted the consensus changes in #2542 so that we can merge them ASAP, avoiding merge conflicts.

nc6 reviewed (changes requested):
I think we need to work out the correct behaviour for `fail`/`error` before making this change. At the moment the semantics change somewhat.

mrBliss commented:
This includes https://github.com/input-output-hk/ouroboros-network/pull/2547 until that is merged (if this PR is merged first, then I'll close the other one).

I think I have fixed the failing test (let's see what CI says). `fail` was used in `m`, not `STM m`, so I just switched back to `fail` and added a `MonadFail` constraint.

Using `fail` in `STM m` would require more changes: we shouldn't add a `MonadFail` instance, because the real `STM` monad doesn't have such an instance. It's fine to have a `MonadThrow (STM m)` constraint, so we could switch some uses of `fail`/`error` to `throwM` (+ `userError` or something custom?). But I'd stick to what this PR does, i.e., use `error`, if possible. It appears to be mostly used for impossible situations, in which case `error` makes sense.

mrBliss commented:
I think the [test failure in CI](https://hydra.iohk.io/build/3952014/nixlog/1) could be related to a timing issue?
```
      prop_timeouts_io withTimeoutSerialAlternative:  FAIL (165.75s)
        *** Failed! Falsified (after 100 tests and 16 shrinks):
        [(0s,0.005s),(0s,0.005s),(0s,0s),(0.002s,0.005s),(0s,0.001s),(0s,0s),(0.008s,0.003s),(0s,0.004s),(0.001s,0.009s),(0.004s,0.003s),(0.001s,0.003s),(0s,0.005s),(0.001s,0.005s),(0.005s,0.003s),(0.004s,0.004s),(0.005s,0.004s),(0.002s,0.004s),(0s,0.003s),(0.005s,0.004s),(0s,0.005s),(0.005s,0.009s),(0.004s,0s),(0.008s,0.009s),(0.003s,0.003s),(0.002s,0s),(0.004s,0.001s),(0.006s,0.01s)]
        failure on timeout test #13
        intendedTimeoutDuration: 0.001s
        intendedActionDuration:  0.005s
        intendedOverallDuration: 0.001s
        actualOverallDuration:   0.113486s
        positiveSchedulingTolerance: 0.1s
        violation of timer sanity property:
          actualOverallDuration <= intendedOverallDuration
                                 + positiveSchedulingTolerance
        Use --quickcheck-replay=851330 to reproduce.
```

coot commented:
@mrBliss these might be valid p2p-governor test failures, I'll open a ticket for them.  Do you also see it locally (I suspect these failing cases are are rare.)?

coot commented on typed-protocols-examples/src/Network/TypedProtocol/Channel.hs:
```diff
@@ -184,7 +183,7 @@ createPipelineTestChannels sz = do
         send x = atomically $ do
                    full <- isFullTBQueue bufferWrite
-                   if full then fail failureMsg
+                   if full then error failureMsg
```
Both `STM SimM` and `STM IO` have `MonadThrow` instances.  It was designed this way so `MonadThrow` has to be declared explicitly.

coot commented:
> I think the test failure in CI could be related to a timing issue?

Which platform is that? I've only seen it on Windows so far.

coot commented on network-mux/src/Network/Mux.hs:
```diff
@@ -524,8 +524,8 @@ runMiniProtocol Mux { muxMiniProtocols, muxControlCmdQueue , muxStatus}
       unless (status == StatusIdle) $
-        fail $ "runMiniProtocol: protocol thread already running for "
-            ++ show ptclNum ++ " " ++ show ptclDir'
+        error $ "runMiniProtocol: protocol thread already running for "
```
For `SimM` both do the same thing:
```
instance MonadFail (STM s) where
  fail msg = STM $ \_ -> ThrowStm (toException (ErrorCall msg))

instance MonadThrow (STM s) where
  throwM e = STM $ \_ -> ThrowStm (toException e)
```

coot reviewed (approved):
LGTM, I think there shouldn't be too many conflicts with `coot/connection-manager` (both modify `PeerSelection` modules).

mrBliss commented:
> @mrBliss these might be valid p2p-governor test failures, I'll open a ticket for them. Do you also see it locally (I suspect these failing cases are are rare.)?

I can't reproduce it, even with the same `quickcheck-replay`.

Am I correct in assuming this test relies on actual (IO) timing? If so, then I suggest relaxing the bound a bit, because it will probably fail whenever the CI build machine is overloaded. We had a test like that in consensus that failed every few weeks until we finally relaxed its time constraints.

mrBliss commented:
bors merge

mrBliss commented:
Uggh, the cached failure made the PR fail :slightly_frowning_face:

mrBliss commented:
bors merge
